### ISSUE TITLE
[BUG] AnimatePresence exit animations not working with createPortal

### ISSUE THREAD

**Issue by winkcor** [framer/motion#2692]
**1. Read the [FAQs](#faqs) 👇**

**2. Describe the bug**

AnimatePresence only work when it send as children to createPortal it not effected when its as parent of createPortal.

**3. IMPORTANT: Provide a CodeSandbox reproduction of the bug**

A [CodeSandbox](https://codesandbox.io/p/devbox/wizardly-volhard-xfxnhf?file=%2Fapp%2Fpage.tsx%3A76%2C46)

**4. Steps to reproduce**
this is not work and element not render
```tsx
<AnimatePresence>
  {isOpen &&
    createPortal(
    <>...</>,
      document.body
    )}
</AnimatePresence>
```
in below code element render but exit animation not working
```tsx

return(
  createPortal(
          <AnimatePresence>
              {isOpen && <>...</>}
          </AnimatePresence>,
        document.body
      )
);
```
**5. Expected behavior**

AnimatePresence and Exit animation work for createPrtal as regular JSX element.

**7. Environment details**

    Chrome 125.0.6422.78
    Firefox 126.0.1
    Windows 10

**Comment by june21x** [comment]
> in below code element render but exit animation not working
> ```tsx
> return(
>  createPortal(
>        <AnimatePresence>
>             {isOpen && <>...</>}
>        </AnimatePresence>,
>       document.body
>    )
> );
>```

Hi @winkcor, I've tried out this solution, it seems fine to me. [CodeSandbox](https://codesandbox.io/p/devbox/kind-leaf-h2p73n?file=%2Fapp%2Fpage.tsx%3A83%2C7-101%2C8)


https://github.com/framer/motion/assets/35992041/ff4c1b0a-ba3d-46ef-8277-3b1b81099527


Environments:
```
Chrome 125.0.6422.114 (Official Build) (arm64)
macOS Sonoma 14.5
```

**Comment by winkcor** [comment]
> in below code element render but exit animation not working
> ```tsx
> return(
>  createPortal(
>        <AnimatePresence>
>             {isOpen && <>...</>}
>        </AnimatePresence>,
>       document.body
>    )
> );
> ```

 Hi @june21x, you are right, this code works, but this is a server component, if we do not set the isOpen condition before `createPortal`, the component will be rendered on the server side, and the `document.body` is not defined on the server, and if you look at the server console, you will see the following error:
 ```tsx
 ⨯ src\components\modal\index.tsx (51:15) @ document
 ⨯ ReferenceError: document is not defined
    at ModalContainer.Modal (./src/components/modal/index.tsx:81:23)
digest: "3267920918"
  49 |                 }
  50 |             </AnimatePresence>
>51 |             , document.body, this.key);
     |               ^
  52 |
  53 |     };
  54 | }
  ```
  so we have to use the following code, but with this code, the component is not rendered
>  ```tsx
><AnimatePresence>
>  {isOpen &&
>    createPortal(
>    <>...</>,
>      document.body
>    )}
></AnimatePresence>
>  ```

**Comment by june21x** [comment]
@winkcor I see, I have modified my solution to align with your requirement, by wrapping a `Fragment` around `createPortal()`. [CodeSandbox](https://codesandbox.io/p/devbox/kind-leaf-h2p73n?file=%2Fapp%2Fpage.tsx%3A89%2C1)

```tsx
<AnimatePresence>
  {isOpen &&
    <>
      {createPortal(
        <>...</>,
        document.body
      )}
    </>
  }
</AnimatePresence>
```

Based on the source code [here](https://github.com/framer/motion/blob/c5e5175fd0820135ebe22434cf13e9a11a43b74d/packages/framer-motion/src/components/AnimatePresence/index.tsx#L39),

https://github.com/framer/motion/blob/c5e5175fd0820135ebe22434cf13e9a11a43b74d/packages/framer-motion/src/components/AnimatePresence/index.tsx#L39

My guess is that `<AnimatePresence/>` doesn't detect `ReactPortal` as its children because somehow `ReactPortal` is not a valid `ReactElement`(even though `ReactPortal` does extends `ReactElement`) when validated with `isValidElement()`. That's why it's not rendering.

```tsx
console.log(isValidElement(createPortal(<></>, document.body)})); // false

console.log(isValidElement(<>{createPortal(<></>, document.body)}</>)); // true
```

**Comment by mattgperry** [comment]
I suspect this is right - ideally the child should also have a unique `key` which `createPortal` can't have

**Comment by YaminZheng** [comment]
`import { createPortal } from "react-dom";
import { styleMap, animationMap } from "./config";
import { AnimatePresence, motion } from "framer-motion";

type Direction = "btt" | "ttb" | "rtl" | "ltr";
export type DrawerProps<P = {}> = ComponentProps<
  { isOpen: boolean; setIsOpen: (v: boolean) => any; direction?: Direction } & P
>;

export default function ({ isOpen, setIsOpen, children, direction = "btt", className, ...otherProps }: DrawerProps) {
  const key = useRef(Date.now());
  const initial = styleMap[direction];
  const animate = animationMap[direction];
  const [isShowChild, setIsShowChild] = useState(false);

  const parentVariants = {
    open: { opacity: 1, transition: { duration: 0.2 } },
    closed: { opacity: 0, transition: { duration: 0.2 } },
  };

  const childVariants = { hidden: initial, visible: animate };

  return (
    <AnimatePresence mode="wait">
      {isOpen &&
        createPortal(
          <>
            <motion.div
              key={`drawer-${key.current}`}
              className="fixed inset-0 bg-[#333] bg-opacity-60 z-999"
              initial="closed"
              exit="closed"
              animate="open"
              variants={parentVariants}
              onTouchEnd={stopPropagation(() => setIsShowChild(false))}
              onAnimationComplete={(definition) => {
                setIsShowChild(definition === "open");
              }}
            >
              <AnimatePresence mode="wait">
                {isShowChild && (
                  <motion.div
                    onTouchEnd={stopPropagation()}
                    key={`child-${key.current}`}
                    initial="hidden"
                    animate="visible"
                    exit="hidden"
                    variants={childVariants}
                    className={classNames("absolute", className)}
                    onAnimationComplete={(definition) => {
                      setIsOpen(definition === "visible");
                    }}
                    {...otherProps}
                  >
                    {children}
                  </motion.div>
                )}
              </AnimatePresence>
            </motion.div>
          </>,
          document.body
        )}
    </AnimatePresence>
  );
}`
This is a Drawer component, I don't understand what the problem is, if there is no createPortal package, the animation is complete

---
"framer-motion": "^11.15.0",
"react": "^18.3.1",
"react-dom": "^18.3.1",
---

**Comment by YaminZheng** [comment]
```tsx
import { createPortal } from "react-dom";
import { styleMap, animationMap } from "./config";
import { AnimatePresence, motion } from "framer-motion";

type Direction = "btt" | "ttb" | "rtl" | "ltr";
export type DrawerProps<P = {}> = ComponentProps<
  { isOpen: boolean; setIsOpen: (v: boolean) => any; direction?: Direction } & P
>;

export default function ({ isOpen, setIsOpen, children, direction = "btt", className, ...otherProps }: DrawerProps) {
  const key = useRef(Date.now());
  const initial = styleMap[direction];
  const animate = animationMap[direction];
  const [isShowChild, setIsShowChild] = useState(false);

  const parentVariants = {
    open: { opacity: 1, transition: { duration: 0.2 } },
    closed: { opacity: 0, transition: { duration: 0.2 } },
  };

  const childVariants = { hidden: initial, visible: animate };

  return (
    <AnimatePresence mode="wait">
      {isOpen &&
        createPortal(
          <>
            <motion.div
              key={`drawer-${key.current}`}
              className="fixed inset-0 bg-[#333] bg-opacity-60 z-999"
              initial="closed"
              exit="closed"
              animate="open"
              variants={parentVariants}
              onTouchEnd={stopPropagation(() => setIsShowChild(false))}
              onAnimationComplete={(definition) => {
                setIsShowChild(definition === "open");
              }}
            >
              <AnimatePresence mode="wait">
                {isShowChild && (
                  <motion.div
                    onTouchEnd={stopPropagation()}
                    key={`child-${key.current}`}
                    initial="hidden"
                    animate="visible"
                    exit="hidden"
                    variants={childVariants}
                    className={classNames("absolute", className)}
                    onAnimationComplete={(definition) => {
                      setIsOpen(definition === "visible");
                    }}
                    {...otherProps}
                  >
                    {children}
                  </motion.div>
                )}
              </AnimatePresence>
            </motion.div>
          </>,
          document.body
        )}
    </AnimatePresence>
  );
}
```
This is a Drawer component, I don't understand what the problem is, if there is no createPortal package, the animation is complete
@mattgperry
